pkg/controllers/clusterresourceplacement/clusterresourceplacement_controller.go

```diff
@@ -198,6 +298,63 @@ func (r *Reconciler) lookupLatestClusterPolicySnapshot(ctx context.Context, crp
 	return &snapshotList.Items[index], lastPolicyIndex, nil
 }
 
+// lookupLatestResourceSnapshot finds the latest snapshots and its resource index.
+// There will be only one active resource snapshot if exists.
+// It first checks whether there is an active resource snapshot.
+// If not, it finds the one whose resourceIndex label is the largest.
+// The resource index will always start from 0.
+// Return error when 1) cannot list the snapshots 2) there are more than one active resource snapshots 3) snapshot has the
+// invalid label value.
+// 2 & 3 should never happen.
+func (r *Reconciler) lookupLatestResourceSnapshot(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) (*fleetv1beta1.ClusterResourceSnapshot, int, error) {
+	snapshotList := &fleetv1beta1.ClusterResourceSnapshotList{}
+	latestSnapshotLabelMatcher := client.MatchingLabels{
+		fleetv1beta1.CRPTrackingLabel:      crp.Name,
+		fleetv1beta1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+	}
+	crpKObj := klog.KObj(crp)
+	if err := r.Client.List(ctx, snapshotList, latestSnapshotLabelMatcher); err != nil {
+		klog.ErrorS(err, "Failed to list active clusterResourceSnapshots", "clusterResourcePlacement", crpKObj)
+		return nil, -1, controller.NewAPIServerError(false, err)
+	}
+	if len(snapshotList.Items) == 1 {
+		resourceIndex, err := parseResourceIndexFromLabel(&snapshotList.Items[0])
+		if err != nil {
+			klog.ErrorS(err, "Failed to parse the resource index label", "clusterResourceSnapshot", klog.KObj(&snapshotList.Items[0]))
+			return nil, -1, controller.NewUnexpectedBehaviorError(err)
+		}
+		return &snapshotList.Items[0], resourceIndex, nil
+	} else if len(snapshotList.Items) > 1 {
+		// It means there are multiple active snapshots and should never happen.
+		err := fmt.Errorf("there are %d active clusterResourceSnapshots owned by clusterResourcePlacement %v", len(snapshotList.Items), crp.Name)
+		klog.ErrorS(err, "It should never happen", "clusterResourcePlacement", crpKObj)
```

Comment:
Maybe it's just me but I don't think "It should never happen" is a valid error message.